### PR TITLE
[MultiDB]: install default db config file when installing libswsscommon if there is no such a file

### PR DESCRIPTION
**Pull request by dzhangalibaba**
* distribute  database_config.json into /var/run/redis/sonic-db/ when installing libswsscommon.deb in postinst script

Signed-off-by: Dong Zhang d.zhang@alibaba-inc.com

### CODE REVIEW

**Comment by dzhangalibaba** [comment]
@qiluo-msft installing default db config file if there is no one in expected location.

**Comment by qiluo-msft** [comment]
> 

Thanks for implement the logic manually. I guess [http://man7.org/linux/man-pages/man5/deb-conffiles.5.html](http://man7.org/linux/man-pages/man5/deb-conffiles.5.html) is just for that, could you help explorer?

---
Refers to: debian/postinst:12 in ef01c58. [](commit_id = ef01c58e8793c4dcee4bef4b6402fc52963654bf, deletion_comment = False)

**Comment by dzhangalibaba** [comment]
> > 
> 
> Thanks for implement the logic manually. I guess http://man7.org/linux/man-pages/man5/deb-conffiles.5.html is just for that, could you help explorer?
> 
> Refers to: debian/postinst:12 in [ef01c58](https://github.com/Azure/sonic-swss-common/commit/ef01c58e8793c4dcee4bef4b6402fc52963654bf). [](commit_id = [ef01c58](https://github.com/Azure/sonic-swss-common/commit/ef01c58e8793c4dcee4bef4b6402fc52963654bf), deletion_comment = False)

Hi Qi, conffiles works, but there is one concern. When we use conffiles, if the config file already exists, installing deb pkg will give a prompt as below. The user needs to decide what to do next.
Did somne search, looks there is no better to skip this prompt in deb pkg installing. If we use this way,  we need to add options when issusing "sudo dpkg -i --force-confold libswsscommon_1.0.0_amd64.deb" with --force-confold option to skip the prompt. Which means we need to modify all the build script which is using "dpkg -i libswsscommon_1.0.0_amd64.deb". This seems not good to me, it introduce more changes and did not get any big benefits. It also force installing to not overwrite existing files, which may be not good in future when there is a need to overwrite. Using manually copy in postinst can avoid this problem and didn't introduce extra changes.


```
Configuration file '/var/run/redis/sonic-db/database_config.json'
 ==> File on system created by you or by a script.
 ==> File also in package provided by package maintainer.
   What would you like to do about it ?  Your options are:
    Y or I  : install the package maintainer's version
    N or O  : keep your currently-installed version
      D     : show the differences between the versions
      Z     : start a shell to examine the situation
 The default action is to keep your current version.
*** database_config.json (Y/I/N/O/D/Z) [default=N] ? Y
```

**Comment by qiluo-msft** [comment]
Thanks @dzhangalibaba !

Could you please share your code about "conffiles works" in another PR? I would like to explorer further. To suppress dpkg prompt is the right thing to do in future.

**Comment by dzhangalibaba** [comment]
> Thanks @dzhangalibaba !
> 
> Could you please share your code about "conffiles works" in another PR? I would like to explorer further. To suppress dpkg prompt is the right thing to do in future.

FYI, https://github.com/Azure/sonic-swss-common/pull/323

File: debian/postinst [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# This `DEBIAN/postinst` script is run post-installation
+DST_PATH="/var/run/redis/sonic-db"
+DST_FILE="database_config.json"
+SRC_PATH="/usr/share/swss"
+
+# if there is no $DST_FILE, it is needed to copy one default to $DST_FILE
```
Comment (kcudnik) [inline debian/postinst]: if file exists we should copy this one one wih suffix ".new" to the same directory to indicate that there is new configuration, btw. i though that there is mechanism already in debian packages to do that, instead manually do that for each file, here is some nice explanation: https://raphaelhertzog.com/2010/09/21/debian-conffile-configuration-file-managed-by-dpkg/

**Comment by kcudnik** [comment]
Just as a question, our future goal is  to support multiple syncd's running across system, controlled by single OA agent, some requirements i posted in my long email:
* each syncd must have global unique id (0..255 etc)
* each syncd must have database assigned (redis db index)
* each syncd must support multiple switches, and each switch must have defined HWINFO for it (SAI_SWITCH_ATTR_SWITCH_HARDWARE_INFO (and maybe also SAI_SWITCH_ATTR_SWITCH_PROFILE_ID)  this information is needed to uniquely identify every switch instance globally

my question here is whether we want this information in this file or do we need other file ? this switch/syncd info is something more than db config, but it strictly depends on database configuration which is defined in this file

also different syncd instances can ha same database server used, but asic_db index must be different for each syncd instance

File: debian/postinst (hunk 1 above)
Comment (dzhangalibaba) [reply debian/postinst]: We can add suffix to save this conf, but we won’t use confines as explained under Qi’s comment, we don’t want prompt.

**Comment by dzhangalibaba** [comment]
> Just as a question, our future goal is to support multiple syncd's running across system, controlled by single OA agent, some requirements i posted in my long email:
> 
> * each syncd must have global unique id (0..255 etc)
> * each syncd must have database assigned (redis db index)
> * each syncd must support multiple switches, and each switch must have defined HWINFO for it (SAI_SWITCH_ATTR_SWITCH_HARDWARE_INFO (and maybe also SAI_SWITCH_ATTR_SWITCH_PROFILE_ID)  this information is needed to uniquely identify every switch instance globally
> 
> **my question here is whether we want this information in this file or do we need other file?** this switch/syncd info is something more than db config, but it strictly depends on database configuration which is defined in this file
> 
> note: different syncd instances can ha same database server used, but asic_db index must be different for each syncd instance

I think we need extend current conf file to support the case you mention. And then add the backend logic to parse conf file to finish the work what you case want

**Comment by kcudnik** [comment]
seems good to me

File: debian/postinst (hunk 1 above)
Comment (kcudnik) [reply debian/postinst]: ok

**Pull request merged**